### PR TITLE
Swap go-version for blang/semver in catalog version validation

### PR DESCRIPTION
**Pull request by jakefhyde**
Replace `go-version` check with `blang/semver`. 

Elected to use `blang/semver` as opposed to `Masterminds/semver` because the former was already being used in the file, and does handle prerelease correctly (after removing v prefix). Also updated unit tests to include `-0` in min and max versions, as well as add tests for max/min only present.

Related Issue: https://github.com/rancher/rancher/issues/32930

### CODE REVIEW

**Review by paynejacob — DISMISSED** [review]
as long as we are not introducing another semver library.

**Review by paynejacob — APPROVED** [review]

File: pkg/catalog/manager/manager.go [hunk 1]
```diff
@@ -243,24 +241,38 @@ func (m *Manager) ValidateKubeVersion(template *v3.CatalogTemplateVersion, clust
...
+	}
+	constraint, err := semver.ParseRange(versionRange)
+	if err != nil {
+		logrus.Errorf("failed to parse constraint for rancher version %s: %v", versionRange, err)
+		return nil
 	}
 
+	rancherVersion, err := semver.Parse(serverVersion)
+	if err != nil {
+		return err
```
Comment (rmweir) [inline pkg/catalog/manager/manager.go:271]: Seems odd to treat these differently, I think it may make sense to log and swallow both since it is probably more annoying to not be able to install a catalog you should be able to install then it is bad that you can install a catalog that you shouldn't be able to install. I do suppose an invalid upgrade could be kind of bad though. @snasovich thoughts?

File: pkg/catalog/manager/manager_test.go [hunk 2]
```diff
@@ -15,20 +15,20 @@ func TestLatestAvailableTemplateVersion(t *testing.T) {
 				{
 					ExternalID:        "catalog://?catalog=library&template=artifactory-ha&version=0.12.16",
 					Version:           "0.12.16",
-					RancherMinVersion: "v2.2.0",
-					RancherMaxVersion: "v2.3.0",
+					RancherMinVersion: "v2.2.0-0",
```
Comment (rmweir) [inline pkg/catalog/manager/manager_test.go:18]: Can you explain what the "-0" is for? I don't think the catalogs we use would include this in the constraints and I don't see you adding it to the versions in ValidateRancherVersion.

File: tests/integration/suite/test_app.py [hunk 3]
```diff
@@ -377,7 +376,6 @@ def test_app_create_validation(admin_mc, admin_pc, custom_catalog,
         app1 = admin_pc.client.create_app(app_data)
         remove_resource(app1)
     assert e.value.error.status == 422
-    assert e.value.error.message == 'rancher max version exceeded'
```
Comment (rmweir) [inline tests/integration/suite/test_app.py:380]: Why remove these? Are there not new error messages that can be checked?

**Review by rmweir — CHANGES_REQUESTED** [review]

File: pkg/catalog/manager/manager.go (hunk 1 above)
Comment (jakefhyde) [reply pkg/catalog/manager/manager.go:271]: This was how it is being done for the kubernetes version, but tI think that makes sense since the only way it will fail to parse is if the server version has been manually set.

File: tests/integration/suite/test_app.py (hunk 3 above)
Comment (jakefhyde) [reply tests/integration/suite/test_app.py:380]: Good catch. I've decided to add checks for part of the error message, since formatting the error client side was sort of cumbersome

File: pkg/catalog/manager/manager_test.go (hunk 2 above)
Comment (jakefhyde) [reply pkg/catalog/manager/manager_test.go]: Updated, do not need it anymore.

File: pkg/catalog/manager/manager.go (hunk 1 above)
Comment (snasovich) [reply pkg/catalog/manager/manager.go:271]: Yes, it matches the following checks so I thought this inconsistency is OK when I reviewed: https://github.com/rancher/rancher/blob/e013338b5eb1271ced81b7ff7ebae093f664d7dc/pkg/catalog/manager/manager.go#L218-L236

I could see some logic in it - if there is the first error, it's an issue with constraints on `CatalogTemplateVersion` so we err on the side of "let's allow to install it" while the second error is when Rancher version is messed up and we block installs against it.

Since I don't know the history of why error handling is inconsistent and there is not a comment that explains it I find it safer to follow the same logic here too.

**Review by paynejacob — APPROVED** [review]

**Review by snasovich — COMMENTED** [review]
I swear I submitted my comment yesterday but now I see it shows as Pending. Maybe my comment is something we can huddle about with @rmweir @paynejacob.

File: pkg/catalog/manager/manager.go [hunk 4]
```diff
@@ -243,24 +241,38 @@ func (m *Manager) ValidateKubeVersion(template *v3.CatalogTemplateVersion, clust
...
+	serverVersion = strings.TrimPrefix(serverVersion, "v")
 
-	if rancherMax != "" && !mVersion.Compare(serverVersion, rancherMax, "<=") {
-		return fmt.Errorf("rancher max version exceeded")
+	versionRange := ""
+	if template.Spec.RancherMinVersion != "" {
+		versionRange += " >=" + strings.TrimPrefix(template.Spec.RancherMinVersion, "v")
+	}
+	if template.Spec.RancherMaxVersion != "" {
+		versionRange += " <=" + strings.TrimPrefix(template.Spec.RancherMaxVersion, "v")
```
[the remark this answers is not in the document]
Comment (jakefhyde) [reply pkg/catalog/manager/manager.go:258]: Some use `x.y.z-alpha1` and `x.y.99-0` together, like the following:
https://github.com/rancher/system-charts/blob/5a757daa2a82ab4ffeedadbf1ad68d9ec1661524/charts/rancher-logging/0.2.5/questions.yml#L1-L2
whereas others use `x.y.z-rc1` and `x.y.z`, like so:
https://github.com/rancher/system-charts/blob/9d3c12a9704b5501baa87193a07fb9f953c465af/charts/rancher-monitoring/v0.2.1/questions.yml#L1-L2
I think regardless it's safe to say we need more unit and integration tests covering this, because the integration tests at https://github.com/rancher/integration-test-charts only use `x.y.z` for both min and max version. This is also vastly different from the actual catalog charts, as they all use `x.y.z-(alpha/rc)1` (except for one [dynatrace chart](https://github.com/rancher/helm3-charts/blob/136a0a384c31a88393bdc56770ba8c848aa2eac9/charts/dynatrace-operator/v0.2.1/questions.yml#L4)).

Also to answer you're question, yes, blang has no such restriction on prereleases, as the unit tests (not the integration tests) would fail. This is achievable with https://github.com/Masterminds/semver, but would require adding prereleases to the max and min version that do not have it. 

A real concrete example is the following:
```
$ vert ">=2.5.6-rc1,<=2.5.7" 2.5.6-rc1
$ echo $?
1
```
Which comes from the monitoring chart listed above. This works as is right now with blang/semver, which I can only assume is what we want otherwise most of the catalogs would disappear for an rc, but it also works with go-version, so I guess the question is are we going to update all of the charts with min and max version to add `-0` (either in the charts or in the validation check) if there is no prerelease for the given min/max?

[the remark this answers is not in the document]
Comment (snasovich) [reply pkg/catalog/manager/manager.go:258]: @jakefhyde , thank you for detailed explanation and research on what ranges our charts use.
I've played with previous github.com/mcuadros/go-version based code and new github.com/blang/semver based one and found one mismatch, but I think it's actually correct.
Basically in go-version 2.5.8 or 2.5.8-rc1 fits 2.5.7-alpha1 to 2.5.8-0 range while it doesn't in blang's case - https://play.golang.org/p/hPo4E3ZEbnL. The latter is more correct as the intention of <=2.5.8-0 is to essentially make the upper range not include anything from 2.5.8 release line.

Bottom line I think we're fine as far as the actual code, but it would be nice to update at least unit tests to use realistic ranges we're using rather than ambiguous ranges as we use here https://github.com/jakefhyde/rancher/blob/833a37a62556c448c3599573b107696ea2033e52/pkg/catalog/manager/manager_test.go#L15-L32 as for example 2.1.0 fits 2 ranges.

I know it's a bit beyond what the issue is immediately about, but we want to do the right thing here and have unit tests represent the realistic ranges we use.

File: pkg/catalog/manager/manager_test.go [hunk 5]
```diff
@@ -53,17 +53,172 @@ func TestLatestAvailableTemplateVersion(t *testing.T) {
...
+	testLatestAvailableTemplateVersion(t, "v2.0.1", "0.12.14", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.0.2-beta", "0.12.14", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.1.0", "0.12.15", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "v2.2.5", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "dev", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "master", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "master-head", "0.12.16", templateWithoutMaxRancherVersion)
+	testLatestAvailableTemplateVersion(t, "", "0.12.16", templateWithoutMaxRancherVersion)
+
+	testLatestAvailableTemplateVersion(t, "v2.2.0-alpha1", "0.12.16", templateWithMinPrerelease)
```
Comment (snasovich) [inline pkg/catalog/manager/manager_test.go:208]: There is no test case here that expects 0.12.15. However, it doesn't seem possible as the range for 0.12.16 (`[v2.2.0-alpha1, v2.3.0]`) fully includes the range for 0.12.15 (`[v2.2.0-alpha1, v2.2.0]`) so 0.12.15 won't ever be used.

Now, these ranges don't really match what we're generally using per the examples you provided in your comment https://github.com/rancher/rancher/pull/35507/files#r751808450.

`templateWith99MaxPatchVersion` does match one of these examples, but neither `templateWithMinPrerelease` nor `templateWithMinAndMaxPrerelease` match "x.y.z-rc1 to x.y.(z+1)" example. So I would recommend replacing these 2 sets with new set that has ranges like `[2.2.0-rc1, 2.2.1]` and `[2.2.1-rc1, 2.2.2]` and have a few test cases around these versions.

**Review by snasovich — CHANGES_REQUESTED** [review]
I think unit tests need a bit more work per the comment I have submitted.

File: pkg/catalog/manager/manager_test.go (hunk 5 above)
Comment (jakefhyde) [reply pkg/catalog/manager/manager_test.go]: Done, also added more tests to the 99 version to highlight the case for ranges like `>=v2.2.0-alpha1,<=v2.2.99` being followed up by `>=v2.3.0-alpha1,<=v2.3.99` (`v2.3.0-0` should not be accepted by either).

**Review by snasovich — APPROVED** [review]
Updated tests LGTM, but we may still want to hold off merging it for 2.6.3 and may want to wait till 2.6.4.
Asked @cbron 's input on this.

**Review by paynejacob — DISMISSED** [review]

File: pkg/catalog/manager/manager.go (hunk 1 above)
Comment (rmweir) [reply pkg/catalog/manager/manager.go:271]: @snasovich yeah that makes sense.

**Review by rmweir — APPROVED** [review]
LGTM

**Comment by snasovich** [comment]
Decided not to merge it for 2.6.3, so converted to draft to prevent accidental merging.

**Review by paynejacob — APPROVED** [review]

**Review by snasovich — APPROVED** [review]
